README.md

# SDMutils <img src="inst/sticker/sdmutils_hex.png" width="180px" align="right"/>

This package provides a set of commonly used functions that are used to
summarise occurrence records in both space and time, identify outliers,
prepare data for SDM input, and write output files from the various SDM
steps in a standardised format.

## Installation

To install from GitHub using the
[remotes](https://github.com/r-lib/remotes) package, run:

`remotes::install_github("DomHenry/SDMUtils")`

## Functions currently available

    ## [1] "check_occ_range.R"    "plot_occ_spatial.R"   "plot_occ_temporal.R" 
    ## [4] "SDMutils-package.R"   "utils-pipe.R"         "write_bart_plots.R"  
    ## [7] "write_bart_rasters.R"


NEWS.md
# SDMutils (development version)

## SDMutils 0.1.0.9006

-   Added five helper functions.

## SDMutils 0.1.0.9000

-   Added a `NEWS.md` file to track changes to the package.


R/check_occ_range.R
#' Check which occurrence points fall outside of known range
#' @description The known range can be either a IUCN range, hand-delineated expert polygon or a QDS shapefile.
#' @param range polygon defining the species range (can be an `sf` or `sp` object)
#' @param occ_data an `sf` spatial data frame with occurrence points
#' @note If a point is located beyond the 20km buffer then all of `core`, `buff1`, `buff2`, and `buff3` will be given the value `FALSE`.
#' @author Dominic Henry
#' @details The function creates three buffers around the range and identifies where a point is located based on the buffers (i.e., within the core range, or within a 5km, 10km, or 20km buffer). Four columns are added to the input data frame (`core`, `buff1`, `buff2`, and `buff3`) and are given a logical value depending on where the point is located (`TRUE` if the point is in the core or buffers, `FALSE` if it isn't).
#' @return The `occ_data` input with four columns added (`core`, `buff1`, `buff2`, and `buff3`)
#' @importFrom methods as
#' @importFrom magrittr "%>%"
#' @export
#'
#' @examples
#' \dontrun{
#' data <- check_occ_range(range, occ_data)
#' }
check_occ_range <- function(range, occ_data) {

  range <- as(range, "Spatial")
  buff1 <- raster::buffer(range, width = 0.05)
  buff2 <- raster::buffer(range, width = 0.1)
  buff3 <- raster::buffer(range, width = 0.2)

  b1 <- buff1 - range
  b2 <- buff2 - buff1
  b3 <- buff3 - buff2

  if (nrow(occ_data) == 0) {
    occ_data <- NULL
  } else {
    occ_data <- occ_data %>%
      sf::st_transform(sf::st_crs(range))

    occ_data <- occ_data %>%
      dplyr:: mutate(core = as.logical(sf::st_intersects(occ_data,sf::st_union(sf::st_as_sf(range)), sparse = FALSE))) %>%
      dplyr::mutate(buff1 = as.logical(sf::st_intersects(occ_data, sf::st_as_sf(b1), sparse = FALSE))) %>%
      dplyr::mutate(buff2 = as.logical(sf::st_intersects(occ_data, sf::st_as_sf(b2), sparse = FALSE))) %>%
      dplyr::mutate(buff3 = as.logical(sf::st_intersects(occ_data, sf::st_as_sf(b3), sparse = FALSE)))
  }

  return(occ_data)
}


R/check_occ_expert.R
#' Check which occurrence points fall outside of known range
#' @description The known range can be either a IUCN range, hand-delineated expert polygon or QDS cells.
#' @param range polygon defining the species range (can be an `sf` or `sp` object)
#' @param occ_data an `sf` spatial data frame with occurrence points
#' @note It's important to note a couple of things. The buffers are set at 5, 10, and 20 km
#' @author Dominic Henry
#' @details Explain what the function does
#' @return The `occ_data` input with four columns added (core:buff3)
#' @importFrom methods as
#' @importFrom magrittr "%>%"
#' @export
#'
#' @examples
#' \dontrun{
#' data <- check_occ_expert(range, occ_data)
#' }
check_occ_expert <- function(range, occ_data) {

  # range <- st_union(range)
  # range <- sf::st_make_valid(range)

  range <- as(range, "Spatial")
  buff1 <- raster::buffer(range, width = 0.05)
  buff2 <- raster::buffer(range, width = 0.1)
  buff3 <- raster::buffer(range, width = 0.2)

  # rgeos::gIsValid(range, byid = FALSE, reason=TRUE)

  b1 <- buff1 - range
  b2 <- buff2 - buff1
  b3 <- buff3 - buff2

  if (nrow(occ_data) == 0) {
    occ_data <- NULL
  } else {
    occ_data <- occ_data %>%
      sf::st_transform(sf::st_crs(range))

    occ_data <- occ_data %>%
      dplyr:: mutate(core = as.logical(sf::st_intersects(occ_data,sf::st_union(sf::st_as_sf(range)), sparse = FALSE))) %>%
      dplyr::mutate(buff1 = as.logical(sf::st_intersects(occ_data, sf::st_as_sf(b1), sparse = FALSE))) %>%
      dplyr::mutate(buff2 = as.logical(sf::st_intersects(occ_data, sf::st_as_sf(b2), sparse = FALSE))) %>%
      dplyr::mutate(buff3 = as.logical(sf::st_intersects(occ_data, sf::st_as_sf(b3), sparse = FALSE)))
  }

  return(occ_data)
}


R/plot_occ_spatial.R
#' Create temporal plots of species occurrence data
#' @description This function summarises occurrence data in a spatial context.
#' @param data `sf` point object with species occurrence data
#' @param range `sf` polygon defining the species range (IUCN, expert polygon, or QDS range)
#' @param za shapefile of South Africa read in as an `sf` object
#' @param ptheme ggplot theme object
#' @author Dominic Henry
#' @details The function draws three plots showing occurrence records overlaid on South Africa. Plots are written to the species directory.
#' @note The size of the text and other elements of the plots can be change by customising the `theme` object.
#' @return Spatial JPEG plots: 1) `occ_map_{sppselect}`; 2) `occ_map_decadal_{sppselect}`; and 3) `occ_map_inset_{sppselect}`
#' @return
#' @export
#'
#' @examples
#' \dontrun{
#' plot_occ_spatial(occ_data_sf, range, za, ptheme)
#' }
plot_occ_spatial <- function(data, range, za, ptheme) {

  occ_bbox <- sf::st_make_grid(sf::st_buffer(data, 0.3), n = 1) # Make a grid with one cell

  p1 <- ggplot2::ggplot() +
    ggplot2::geom_sf(data = za, fill = ggplot2::alpha("grey", 0.3), size = 0.5) +
    ggplot2::geom_sf(data = range, fill = NA, col = "dodgerblue", size = 0.8) +
    ggplot2::geom_sf(data = data, size = 1.5, col = "black") +
    ggplot2::theme_bw() +
    ggplot2::ggtitle(glue::glue("{sppselect}: {nrow(data)} occurrence points")) +
    ptheme

  ggplot2::ggsave(glue::glue("{sdm_dir}/{sppselect}/occ_map_{sppselect}.jpg"), plot = p1, height = 12, width = 16)

  p2 <- ggplot2::ggplot() +
    ggplot2::geom_sf(data = za, fill = NA, size = 0.5) +
    ggplot2::geom_sf(data = data, ggplot2::aes(color = decade), size = 3) +
    ggplot2::scale_color_viridis_d(na.value = ggplot2::alpha("grey", 0.7), direction = 1) +
    ggplot2::guides(color = ggplot2::guide_legend(
      title = "Decade",
      override.aes = list(size = 8)
    )) +
    ggplot2::coord_sf(
      xlim = c(sf::st_bbox(occ_bbox)$xmin, sf::st_bbox(occ_bbox)$xmax),
      ylim = c(sf::st_bbox(occ_bbox)$ymin, sf::st_bbox(occ_bbox)$ymax)
    ) +
    ggplot2::theme_bw() +
    ptheme +
    ggplot2::ggtitle(glue::glue("{sppselect}"))

  ggplot2::ggsave(glue::glue("{sdm_dir}/{sppselect}/occ_map_decadal_{sppselect}.jpg"), plot = p2, height = 12, width = 16)

  p3 <- ggplot2::ggplot() +
    ggplot2::geom_sf(data = za, fill = ggplot2::alpha("grey", 0.3), size = 0.5) +
    ggplot2::geom_sf(data = occ_bbox, fill = NA, col = "dodgerblue", size = 1.2) +
    ggplot2::geom_sf(data = data, size = 1.5, col = "black") +
    ggplot2::theme_bw() +
    ggplot2::ggtitle(glue::glue("{sppselect}")) +
    ptheme

  p4 <- ggplot2::ggplot() +
    ggplot2::geom_sf(data = za, fill = ggplot2::alpha("grey", 0.3), size = 0.5) +
    ggplot2::geom_sf(data = data, size = 1.5, col = "black") +
    ggplot2::coord_sf(
      xlim = c(sf::st_bbox(occ_bbox)$xmin, sf::st_bbox(occ_bbox)$xmax),
      ylim = c(sf::st_bbox(occ_bbox)$ymin, sf::st_bbox(occ_bbox)$ymax)
    ) +
    ggplot2::theme_bw() +
    ggplot2::theme(panel.border = ggplot2::element_rect(colour = "dodgerblue", fill = NA, size = 1.6)) +
    ptheme

  p5 <- gridExtra::grid.arrange(grobs = list(p3, p4), nrow = 1)

  ggplot2::ggsave(glue::glue("{sdm_dir}/{sppselect}/occ_map_inset_{sppselect}.jpg"),
                  plot = p5, height = 12, width = 16
  )

  print("SPATIAL PLOTS WRITTEN TO SPECIES SDM OUTPUT DIRECTORY")
}
if(getRversion() >= "2.15.1")  utils::globalVariables(c("sdm_dir", "sppselect", "decade"))


R/plot_occ_temporal.R
#' Create temporal plots of species occurrence data
#' @description This function summarises occurrence data in a temporal context.
#' @param data `sf` point object with species occurrence data
#' @param ptheme ggplot theme
#' @author Dominic Henry
#' @details The function draws two plots showing occurrence record frequencies and writes them to the species directory.
#' @note The size of the text and other elements of the plots can be change by customising the `theme` object. Also take note of the undeclared variables: `sdm_dir`, `sppselect`, `year_date`, and `undated`.
#' @return Temporal JPEG plots: 1) `occ_frequency_date_1980_{sppselect}` and 2) `occ_frequency_date_all_{sppselect}`.
#' @export
#'
#' @examples
#' \dontrun{
#' plot_occ_temporal(occ_data_sf, ptheme)
#' }
plot_occ_temporal <- function(data, ptheme){

  p1 <- data %>%
  ggplot2::ggplot(ggplot2::aes(x = year_date)) +
  ggplot2::geom_bar() +
  ggplot2::ylab("Count of occurrence records") +
  ggplot2::xlab("") +
  ggplot2::labs(
    title = sppselect,
    subtitle = glue::glue("{undated} undated records from a total of {nrow(data)}")
  ) +
  ptheme

ggplot2::ggsave(
  plot = p1,
  filename = glue::glue("{sdm_dir}/{sppselect}/occ_frequency_date_all_{sppselect}.jpg"),
  height = 9, width = 12
)

p2 <- data %>%
  dplyr::filter(year_date > "1980-01-01") %>%
  ggplot2::ggplot(ggplot2::aes(x = year_date)) +
  ggplot2::geom_bar() +
  ggplot2::ylab("Count of occurrence records") +
  ggplot2::xlab("") +
  ggplot2::labs(
    title = sppselect,
    subtitle = glue::glue("{undated} undated records")
  ) +
  ggplot2::scale_x_date(date_breaks = "3 year", date_labels = "%Y") +
  ptheme

ggplot2::ggsave(
  plot = p2,
  filename = glue::glue("{sdm_dir}/{sppselect}/occ_frequency_date_1980_{sppselect}.jpg"),
  height = 9, width = 12
)

print("TEMPORAL PLOTS WRITTEN TO SPECIES SDM OUTPUT DIRECTORY")

}
if(getRversion() >= "2.15.1")  utils::globalVariables(c("sdm_dir", "sppselect",
                                                        "year_date", "undated"))



R/write_bart_rasters.R
#' Write BART raster predictions to GeoTIFF format
#'
#' @description This function writes GeoTIFFs of BART predictions.
#' @param data a `RasterStack` object returned by `embarcadero::predict2.bart()`
#' @author Dominic Henry
#' @details The function write outputs of both binary and probability surfaces at mean values and 95% credible intervals.
#' @note Undeclared variables: `BART_dir`, `sppselect` and `tss_threshold`
#' @return Seven raster layers:
#' \enumerate{
#'  \item `SDM_prob_mean_{sppselect}`: Mean probability of habitat suitability
#'  \item `SDM_prob_lower_{sppselect}`: Lower 95% credible interval probability of habitat suitability
#'  \item `SDM_prob_upper_{sppselect}`: Upper 95% credible interval probability of habitat suitability
#'  \item `SDM_bin_mean_{sppselect}`: Mean binary presence/absence prediction (based on TSS threshold)
#'  \item `SDM_bin_lower_{sppselect}`: Lower 95% credible interval binary presence/absence
#'  \item `SDM_bin_upper_{sppselect}`: Upper 95% credible interval binary presence/absence
#'  \item `SDM_cred_int_width_{sppselect}`: Credible interval probability width (i.e., Upper CI - Lower CI)
#'
#' }
#' @export
#'
#' @examples
#' \dontrun{
#' write_bart_rasters(map)
#' }
write_bart_rasters <- function(data){

  raster::writeRaster(data[[1]],
                      glue::glue("{BART_dir}/SDM_prob_mean_{sppselect}.tif"),
                      format = "GTiff",overwrite = TRUE)

  raster::writeRaster(data[[2]],
                      glue::glue("{BART_dir}/SDM_prob_lower_{sppselect}.tif"),
                      format = "GTiff",overwrite = TRUE)

  raster::writeRaster(data[[3]],
                      glue::glue("{BART_dir}/SDM_prob_upper_{sppselect}.tif"),
                      format = "GTiff",overwrite = TRUE)

  raster::writeRaster(data[[1]] > tss_threshold,
                      glue::glue("{BART_dir}/SDM_bin_mean_{sppselect}.tif"),
                      format = "GTiff",overwrite = TRUE)

  raster::writeRaster(data[[2]] > tss_threshold,
                      glue::glue("{BART_dir}/SDM_bin_lower_{sppselect}.tif"),
                      format = "GTiff",overwrite = TRUE)

  raster::writeRaster(data[[3]] > tss_threshold,
                      glue::glue("{BART_dir}/SDM_bin_upper_{sppselect}.tif"),
                      format = "GTiff",overwrite = TRUE)

  raster::writeRaster(data[[3]]- data[[2]],
                      glue::glue("{BART_dir}/SDM_cred_int_width_{sppselect}.tif"),
                      format = "GTiff",overwrite = TRUE)

  print("RASTERS WRITTEN TO SPECIES DIRECTORY")
}
if(getRversion() >= "2.15.1")  utils::globalVariables(c("BART_dir", "sppselect", "tss_threshold"))


R/write_bart_plots.R
#' Generate maps of BART SDM predictions
#' @description This function writes plots using a combination of PDF and PNG outputs.
#' @param data a `RasterStack` object returned by `embarcadero::predict2.bart()`
#' @param mean_theme a `rasterTheme` used for plotting the mean probability SDM map
#' @param ci_theme a `rasterTheme` used for plotting the posterior width (95% credible interval) SDM map
#' @author Dominic Henry
#' @details The function writes 8 plots related to the species SDM predictions, both probability and binary surfaces. The species' occurrence points have been overlaid on certain plots. Plots are written to the species directory.
#' @note There are several undeclared variables to take note of: `BART_dir`, `tss_threshold`, `occ`, `num_occ_points`, and `occurence_points`
#' @return The following plot files:
#' \enumerate{
#'  \item `SDM_mean_probability.pdf`
#'  \item `SDM_mean_probability.png`
#'  \item `SDM_posterior_width.pdf`
#'  \item `SDM_posterior_width.png`
#'  \item `SDM_four_panel_probability.pdf`
#'  \item `SDM_four_panel_binary.pdf`
#'  \item `SDM_mean_binary.pdf`
#'  \item `SDM_mean_binary.png`
#'
#' }
#' @export
#'
#' @examples
#' \dontrun{
#' write_bart_plots(data = map, mean_theme = maptheme_mean, ci_theme = maptheme_ci)
#' }
write_bart_plots <- function(data, mean_theme, ci_theme) {
  p <- rasterVis::levelplot(data[[1]],
                            maxpixels = 1e10,
                            margin = FALSE,
                            par.settings = mean_theme,
                            scales = list(
                              x = list(draw = FALSE),
                              y = list(draw = FALSE)
                            ),
                            zlim = c(0, 1)
  )

  p1 <- p +
    latticeExtra::layer(sp::sp.points(occurence_points, pch = 20, col = "black", cex = 0.8),
                        data = list(occurence_points = sp::SpatialPoints(occ[1:num_occ_points, ]))
    )

  p2 <- p +
    latticeExtra::layer(sp::sp.points(occurence_points, pch = 20, col = "black", cex = 4),
                        data = list(occurence_points = sp::SpatialPoints(occ[1:num_occ_points, ]))
    )

  pdf(glue::glue("{BART_dir}/SDM_mean_probability.pdf"), width = 16, height = 9)
  print(p1)
  grid::grid.text("Probability of presence",
                  rot = 90,
                  y = grid::unit(0.5, "npc"),
                  x = grid::unit(0.925, "npc"),
                  gp = grid::gpar(fontsize = 15)
  )
  dev.off()

  grDevices::png(glue::glue("{BART_dir}/SDM_mean_probability.png"), width = 2800, height = 2000)
  print(p2)
  grid::grid.text("Probability of presence",
                  rot = 90,
                  y = grid::unit(0.5, "npc"),
                  x = grid::unit(0.925, "npc"),
                  gp = grid::gpar(fontsize = 45)
  )
  dev.off()

  p3 <- rasterVis::levelplot(data[[3]] - data[[2]],
                             maxpixels = 1e10,
                             margin = FALSE,
                             par.settings = ci_theme,
                             scales = list(
                               x = list(draw = FALSE),
                               y = list(draw = FALSE)
                             ),
                             zlim = c(0, 1)
  )

  pdf(glue::glue("{BART_dir}/SDM_posterior_width.pdf"), width = 16, height = 9)
  print(p3)
  grid::grid.text("Posterior width",
                  rot = 90,
                  y = grid::unit(0.5, "npc"),
                  x = grid::unit(0.925, "npc"),
                  gp = grid::gpar(fontsize = 15)
  )
  dev.off()


  grDevices::png(glue::glue("{BART_dir}/SDM_posterior_width.png"), width = 2800, height = 2000)
  print(p3)
  grid::grid.text("Posterior width",
                  rot = 90,
                  y = grid::unit(0.5, "npc"),
                  x = grid::unit(0.925, "npc"),
                  gp = grid::gpar(fontsize = 45)
  )
  dev.off()

  ## Four measure SDM plots
  pdf(glue::glue("{BART_dir}/SDM_four_panel_probability.pdf"), width = 16, height = 9)
  par(mfrow = c(2, 2))
  par(mar = c(2, 1, 2, 5))
  raster::plot(data[[1]],
               main = "Posterior mean",
               box = F, axes = F
  )
  raster::plot(data[[2]],
               main = "Lower 95% CI bound",
               box = F, axes = F
  )
  raster::plot(data[[3]],
               main = "Upper 95% CI bound",
               box = F, axes = F
  )
  raster::plot(data[[3]] - data[[2]],
               main = "Credible interval width",
               box = F, axes = F
  )
  dev.off()

  ## Binary predictions
  pdf(glue::glue("{BART_dir}/SDM_four_panel_binary.pdf"), width = 16, height = 9)
  par(mfrow = c(2, 2))
  raster::plot(data[[1]] > tss_threshold,
               main = "Posterior mean",
               box = F, axes = F
  )
  raster::plot(data[[2]] > tss_threshold,
               main = "Lower 95% CI bound",
               box = F, axes = F
  )
  raster::plot(data[[3]] > tss_threshold,
               main = "Upper 95% CI bound",
               box = F, axes = F
  )
  quant <- terra::quantile(terra::values(data[[3]] - data[[2]]),
                           0.75,
                           na.rm = TRUE
  )
  raster::plot((data[[3]] - data[[2]]) > quant,
               box = FALSE,
               axes = FALSE,
               main = "Highest uncertainty zones",
               axis.args = list(at = pretty(0:1), labels = pretty(0:1)),
               legend.args = list(text = "", side = 2, line = 1.3)
  )
  dev.off()


  ## Mean binary
  pdf(glue::glue("{BART_dir}/SDM_mean_binary.pdf"), width = 16, height = 9)
  par(mfrow = c(1, 1))
  raster::plot(data[[1]] > tss_threshold,
               main = "Predicted presence",
               box = F, axes = F
  )
  terra::points(points(occ[1:num_occ_points, ], col = "blue", pch = 20))
  dev.off()

  grDevices::png(glue::glue("{BART_dir}/SDM_mean_binary.png"), width = 2800, height = 2000)
  par(mfrow = c(1, 1))
  raster::plot(data[[1]] > tss_threshold,
               main = "Predicted presence",
               box = F, axes = F
  )
  terra::points(points(occ[1:num_occ_points, ], col = "blue", pch = 20, cex = 3.5))
  dev.off()

  print("PLOTS WRITTEN TO SPECIES DIRECTORY")
}
if(getRversion() >= "2.15.1")  utils::globalVariables(c("BART_dir", "tss_threshold",
                                                        "occ", "num_occ_points",
                                                        "occurence_points"))
